tutorials/hentropy_search.ipynb

```diff
@@ -0,0 +1,454 @@
+    ")\n",
+    "from botorch.models import SingleTaskGP\n",
+    "from botorch.models.transforms.outcome import Standardize\n",
+    "from botorch.test_functions.synthetic import Ackley\n",
+    "from gpytorch.mlls import ExactMarginalLogLikelihood\n",
+    "\n",
+    "SMOKE_TEST = os.environ.get(\"SMOKE_TEST\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this tutorial, we will explore the 2D Ackley function within the range of [-1, 1]. Our objective is to accomplish two tasks: 1) finding the top-2 values, and 2) determining the min-max of the function. To facilitate the evaluation process, we will define the following functions:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def eval_and_plot(func, cfg, qhes, next_x, data, iteration):\n",
+    "    # Quality of the best decision from the current posterior distribution #############\n",
+    "    # Initialize A consistently across fantasies\n",
+    "    A = torch.rand(\n",
+    "        [1, 1, cfg.num_action, cfg.num_dim_action],\n",
+    "        requires_grad=True,\n",
+    "        device=cfg.device,\n",
+    "        dtype=cfg.dtype,\n",
+    "    )\n",
+    "\n",
+    "    # Initialize optimizer\n",
+    "    optimizer = torch.optim.Adam([A], lr=cfg.acq_opt_lr)\n",
+    "    ba_l, ba_u = cfg.bounds_action\n",
+    "\n",
+    "    for i in range(cfg.acq_opt_iter):\n",
+    "        A_ = A.permute(1, 0, 2, 3)\n",
+    "        A_ = torch.sigmoid(A_) * (ba_u - ba_l) + ba_l\n",
+    "        posterior = qhes.model.posterior(A_)\n",
+    "        fantasized_outcome = qhes.action_sampler(posterior)\n",
+    "        # >>> n_fantasy_at_action_pts x n_fantasy_at_design_pts\n",
+    "        # ... x batch_size x num_actions x 1\n",
+    "\n",
+    "        fantasized_outcome = fantasized_outcome.squeeze(dim=-1)\n",
+    "        # >>> n_fantasy_at_action_pts x n_fantasy_at_design_pts\n",
+    "        # ... x batch_size x num_actions\n",
+    "\n",
+    "        losses = -qhes.loss_function(A=A_, Y=fantasized_outcome)\n",
+    "        # >>> n_fantasy_at_design_pts x batch_size\n",
+    "\n",
+    "        loss = losses.mean(dim=0).sum()\n",
+    "        A.grad = torch.autograd.grad(loss, A)[0]\n",
+    "        optimizer.step()\n",
+    "        optimizer.zero_grad()\n",
+    "\n",
+    "        if i % 200 == 0:\n",
+    "            print(f\"Eval optim round: {i}, Loss: {loss.item():.2f}\")\n",
+    "\n",
+    "    A = torch.sigmoid(A) * (ba_u - ba_l) + ba_l\n",
+    "    eval_metric = qhes.loss_function(A=A, Y=fantasized_outcome)\n",
+    "    eval_metric = eval_metric[0, 0].cpu().detach().numpy()\n",
+    "    optimal_action = A[0, 0].cpu().detach().numpy()\n",
+    "    value = qhes.loss_function(A=A, Y=func(A))[0, 0].cpu().detach().numpy()\n",
+    "    data_x = data.x.cpu().detach().numpy()\n",
+    "    next_x = next_x.cpu().detach().numpy()\n",
```

Comment:
can probably get rid of a bunch of the detaches here by using a `no_grad()` context